agent/src/apiclient.rs

```diff
@@ -0,0 +1,359 @@
+const MAX_ATTEMPTS: i8 = 5;
+const UPDATE_API_SLEEP_DURATION: Duration = Duration::from_millis(10000);
+const UPDATE_API_BUSY_STATUSCODE: &str = "423";
+
+/// The module-wide result type.
+pub type Result<T> = std::result::Result<T, apiclient_error::Error>;
+
+/// UpdateState represents four states during system update process
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+pub enum UpdateState {
+    // Idle: Unknow
+    Idle,
+    // Available: available versions system can update to
+    Available,
+    // Staged: processing system update (refresh, prepare, activate)
+    Staged,
+    // Ready: successfully complete update commands, ready to reboot
+    Ready,
+}
+
+/// UpdateCommand represents three commands to update system
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum UpdateCommand {
+    // Refresh: refresh the list of available updates
+    Refresh,
+    // Prepare: request that the update be downloaded and applied to disk
+    Prepare,
+    // Activate: proceed to "activate" the update
+    Activate,
+}
+
+/// CommandStatus represents three status after running update command
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+enum CommandStatus {
+    Success,
+    Failed,
+    Unknown,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct StagedImage {
+    image: Option<UpdateImage>,
+    next_to_boot: bool,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct CommandResult {
+    cmd_type: UpdateCommand,
+    cmd_status: CommandStatus,
+    timestamp: String,
+    exit_status: u32,
+    stderr: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct UpdateImage {
+    arch: String,
+    version: String,
+    variant: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct OsInfo {
+    pub version_id: String,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct UpdateStatus {
+    update_state: UpdateState,
+    available_updates: Vec<String>,
+    chosen_update: Option<UpdateImage>,
+    active_partition: Option<StagedImage>,
+    staging_partition: Option<StagedImage>,
+    most_recent_command: CommandResult,
+}
+
+fn get_raw_args(mut args: Vec<String>) -> Vec<String> {
+    let mut subcommand_args = vec!["raw".to_string(), "-u".to_string()];
+    subcommand_args.append(&mut args);
+
+    return subcommand_args;
+}
+
+/// Apiclient binary has been volume mounted into the agent container, so agent is able to
+/// invoke `/bin apiclient` to interact with the Bottlerocket Update API.
+/// This function helps to invoke apiclient command.
+async fn invoke_apiclient(args: Vec<String>) -> Result<Output> {
+    let mut attempts: i8 = 0;
+    // Retry up to 5 times in case the Update API is busy; Waiting 10 seconds between each attempt.
+    while attempts < MAX_ATTEMPTS {
+        let output = Command::new(API_CLIENT_BIN)
+            .args(&args.clone())
+            .output()
+            .unwrap();
+        // .context(ApiClientCommand { args.clone() })?;
```

Comment:
I think you forgot to remove the comment here? Remember not to call `unwrap` because it could cause a panic :fearful: